### ISSUE TITLE
ACL Find and edit

### ISSUE THREAD

**Issue by newscriptkid** [grelleum/youtube-network-automation#1]
Hi Gregg,

I Need some help with getting the following code to work (im using python 2.7):

1 - Script connects to Switches which are a mix of telnet, ssh using tacacs and local accounts for authentication. 
2 - Check if any of the following ACL's are on the switch (101,102,test) and if so then paste the config for the respective acl found on the switch only. For e.g. switch A has acl 101, 102 and test acl configured, so all 3 acls will need to updated with new entries, then switchB only has acl 102, so only 102 will be updated on switchB. 
3 - to do show acl xxx (101, 102, testacl) post updating and copy this to a file with the switch name (so i have a record of which switch has which acl applied for future reference)

Below is the config that needs to be pasted on the switches and it will be the same for all 3 acl's (101,102&testacl):

(in config t mode)
access-list extended xxx (depending on which acl is found on the switchs) 
109 permit ip 10.10.0.0 0.0.0.255
exit
access-list resequence xxx 10 10 (xxx = acl number, i.e. 101 or 102 or test)
end
show access-list xxx
wr mem
========

I have tried the following code but its not working, i'm sure the code could be smaller but this is my best effort as a beginner.

import getpass
import sys
import telnetlib
from sys import argv
import csv
import ipaddress
import requests
from netmiko import ConnectHandler

switchdata, device_lists, = argv

reader = csv.DictReader(open(device_lists, 'rb'))

device_lists = []

for line in reader:
  device_lists.append(line)

username = raw_input("Username: ")
password = getpass.getpass("Password ")
enablepw = getpass.getpass("Enable Password ")

for device in device_lists:
 print "\n\n-------------\nDevice: {0} \n---------\n".format(device['host'])
 try:
     device['device_type'] = 'cisco_ios_ssh'
     device['username'] = username
     device['password'] = password
     device['secret'] = enablepw
     net_connect = ConnectHandler(**device)

except:
   try:
       device['device_type'] = 'cisco_ios_telnet'
       device['username'] = username
       device['password'] = password
       device['secret'] = enablepw
       net_connect = ConnectHandler(**device)

   except:
      print "Unable to connect!"
      continue
  net_connect.enable()
print "\n"

output = net_connect.send_command("show access-list 101")

with open('acl_101' + device['host'] + '.txt', 'w' ) as acl101:
  for line in output.splitlines():
    if "101" in line:
    output = net_connect.send_command("enable\n")
    output = net_connect.send_command(password + "\n")
    output = net_connect.send_command("config t\n")
    output = net_connect.send_command("ip access-list extended 101\n")
    output = net_connect.send_command("109 permit ip 10.10.0.0 0.0.0.255 any\n")
    output = net_connect.send_command("exit\n")
    output = net_connect.send_command("access-list resequence 101\n")
    output = net_connect.send_command("end\n")
    output = net_connect.send_command("show access-list 101\n")
    output = net_connect.send_command("show hostname\n")
    print " " + line

output = net_connect.send_command("show ip access-list 102")

with open('acl_102' + device['host'] + '.txt', 'w') as acl102:
  for line in output.splitlines():
    if "102" in line:
    output = net_connect.send_command("enable\n")
    output = net_connect.send_command(password + "\n")
    output = net_connect.send_command("config t\n")
    output = net_connect.send_command("ip access-list extended 102\n")
    output = net_connect.send_command("109 permit ip 10.10.0.0 0.0.0.255 any\n")
    output = net_connect.send_command("exit\n")
    output = net_connect.send_command("access-list resequence 102\n")
    output = net_connect.send_command("end\n")
    output = net_connect.send_command("show access-list 102\n")
    output = net_connect.send_command("show hostname\n")

output = net_connect.send_command("show ip access-list test")

with open('testacl' + device['host'] + '.txt', 'w') as testacl:
  for line in output.splitlines():
    if "testacl" in line:
    output = net_connect.send_command("enable\n")
    output = net_connect.send_command(password + "\n")
    output = net_connect.send_command("config t\n")
    output = net_connect.send_command("ip access-list extended test\n")
    output = net_connect.send_command("109 permit ip 10.10.0.0 0.0.0.255 any\n")
    output = net_connect.send_command("exit\n")
    output = net_connect.send_command("access-list resequence test\n")
    output = net_connect.send_command("end\n")
    output = net_connect.send_command("show access-list testacl\n")
    output = net_connect.send_command("show hostname\n")
    print " " + line

Thanks in advance,
Shah

**Comment by grelleum** [comment]
Hi Shah,

A couple of things I would like to touch upon in your code and your question.

As a suggestion, when posting code you should use three backticks to start and end your code block, so that the indentation does not get lost in the posting. (see the following page).
https://help.github.com/articles/creating-and-highlighting-code-blocks/

Also, it would be helpful to let me know what exactly is not working with your code, for example are you getting a traceback error is it just not sending config, or you are not getting output.

I think I might know what it wrong with it, but first I suggest some of the following things:

When you open your CSV file, you should probably use the `with` context manager so that the file gets properly closed.  Here are two examples of how I would suggest re-writing that section of your code.

```
device_lists = []
with open(device_lists, 'rb') as f:
    for line in csv.DictReader(f):
        device_lists.append(line)
```

This second example uses a list comprehension to do the same thing.  I think it is better than the above method, but use whatever you best understand and is most readable for you.

```
with open(device_lists, 'rb') as f:
    device_lists = [line for line in csv.DictReader(f)]
```

Secondly, I suggest creating a function to handle your connection attempts as below:

```
def get_connection(device):
    device['username'] = username
    device['password'] = password
    device['secret'] = enablepw
    device['device_type'] = 'cisco_ios_ssh'
    try:
        net_connect = ConnectHandler(**device)
    except:  # you should be more explicit with what exceptions you are catching
        try:
            device['device_type'] = 'cisco_ios_telnet'
            net_connect = ConnectHandler(**device)
        except:
            print "Unable to connect!"
            net_connect = False
    return net_connect


for device in device_lists:
    print "\n\n-------------\nDevice: {0} \n---------\n".format(device['host'])
    net_connect = get_connection(device)
    if not net_connect:
        continue
```

When using the `netmiko send_command` method, you do not need to add the `\n` ad the end of the command, as `netmiko` will add a `\n` to the end of the command.  
`net_connect.send_command("enable\n")`
should read:
`net_connect.send_command("enable")`
of course, since you earlier told netmiko to go into enable mode, you don't need the enable command when configuring.

I think the problem with your code may be that you keep reusing the "output" variable name that you are depending on in your loop.

Consider doing something like this:
```
    access_lists = net_connect.send_command("show ip access-lists")
    
    if 'access list 101' in access_lists:
        with open('acl_101' + device['host'] + '.txt', 'w' ) as acl101:
            config = [
                'ip access-list extended 101',
                '109 permit ip 10.10.0.0 0.0.0.255 any',
                'exit',
                'access-list resequence 101',
            ]
            output = net_connect.send_config_set(config)
            net_connect.save_config()
            acl101.write(output + '\n')

    if 'access list 102' in access_lists:
```

**Comment by newscriptkid** [comment]
Hi Gregg,
Thank you so much for taking the time out to reply to my issue (honestly i appreciate it alot).

Thanks for the tips and apologies for the inconvenience, will make sure for future will adhere to those guidelines.

Greatly appreciated for the tips you have offered in making the code more simplier, I have applied your suggestions but my code was in Python2.7 and after editing my code with the above its giving me errors such as the print statements need a bracket and when i add the command from your tutorial (from_future...) it then gives me errors that there is no module for the 'request' library and when i delete that it then doesnt recognize netmiko. 

I think I will have to install python3 to run this code, so will do that today.

With my initial issue it was creating blank files for the devices with the acl's on them.

Will test the code on python 3 shortly and feedback the results.

Thank you once again, 
I owe you!
Shah

**Comment by newscriptkid** [comment]
Hi Gregg,
Im struggling to get that code to work in either python versions, when i run the code it gives me one error, then when i correct that then another, so initially errors were related to parenthesis missing on the print statements, then it was request and netmiko module dont exist, then it was net_connect is not defined and now it gives this error:
```
C:\Users\Shahrukh Aziz\PycharmProjects\untitled1>python aclfind.py csv_file.csv
Traceback (most recent call last):
  File "aclfind.py", line 46, in <module>
    access_lists = net_connect.send_command("show ip access-lists")
AttributeError: 'bool' object has no attribute 'send_command'
```

======

When i re-run my original beginner code it gives me the following error: (python2.7)

```
Traceback (most recent call last):
  File "aclfind.py", line 55, in <module>
    output = net_connect.send_command("config t\n")
  File "C:\Python27\lib\site-packages\netmiko\base_connection.py", line 967, in send_command
    search_pattern))
IOError: Search pattern never detected in send_command_expect: R1\#

C:\Program Files\Sublime Text 3>
```
and my original code is below now with the backticks:

```
import getpass
import sys
import telnetlib
from sys import argv
import csv
import ipaddress
import requests
from netmiko import ConnectHandler

switchdata, device_lists, = argv

reader = csv.DictReader(open(device_lists, 'rb'))

device_lists = []

for line in reader:
  device_lists.append(line)

username = raw_input("Username: ")
password = getpass.getpass("Password ")
enablepw = getpass.getpass("Enable Password ")

for device in device_lists:
  print "\n\n-------------\nDevice: {0} \n---------\n".format(device['host'])

  try:
     device['device_type'] = 'cisco_ios_ssh'
     device['username'] = username
     device['password'] = password
     device['secret'] = enablepw
     net_connect = ConnectHandler(**device)

  except:
    try:
       device['device_type'] = 'cisco_ios_telnet'
       device['username'] = username
       device['password'] = password
       device['secret'] = enablepw
       net_connect = ConnectHandler(**device)

    except:
      print "Unable to connect!"
      continue

net_connect.enable()
print "\n"

output = net_connect.send_command("show access-list 101")

with open('acl_101' + device['host'] + '.txt', 'w' ) as acl101:
  for line in output.splitlines():
    if "101" in line:
      output = net_connect.send_command("enable\n")
      output = net_connect.send_command(password + "\n")
      output = net_connect.send_command("config t\n")
      output = net_connect.send_command("ip access-list extended 101\n")
      output = net_connect.send_command("109 permit ip 10.10.0.0 0.0.0.255 any\n")
      output = net_connect.send_command("exit\n")
      output = net_connect.send_command("access-list resequence 101\n")
      output = net_connect.send_command("end\n")
      output = net_connect.send_command("show access-list 101\n")
      output = net_connect.send_command("show hostname\n")
      print " " + line

output = net_connect.send_command("show ip access-list 102")

with open('acl_102' + device['host'] + '.txt', 'w') as acl102:
  for line in output.splitlines():
    if "102" in line:
      output = net_connect.send_command("enable\n")
      output = net_connect.send_command(password + "\n")
      output = net_connect.send_command("config t\n")
      output = net_connect.send_command("ip access-list extended 102\n")
      output = net_connect.send_command("109 permit ip 10.10.0.0 0.0.0.255 any\n")
      output = net_connect.send_command("exit\n")
      output = net_connect.send_command("access-list resequence 102\n")
      output = net_connect.send_command("end\n")
      output = net_connect.send_command("show access-list 102\n")
      output = net_connect.send_command("show hostname\n")

output = net_connect.send_command("show ip access-list test")

with open('testacl' + device['host'] + '.txt', 'w') as testacl:
  for line in output.splitlines():
    if "testacl" in line:
      output = net_connect.send_command("enable\n")
      output = net_connect.send_command(password + "\n")
      output = net_connect.send_command("config t\n")
      output = net_connect.send_command("ip access-list extended test\n")
      output = net_connect.send_command("109 permit ip 10.10.0.0 0.0.0.255 any\n")
      output = net_connect.send_command("exit\n")
      output = net_connect.send_command("access-list resequence test\n")
      output = net_connect.send_command("end\n")
      output = net_connect.send_command("show access-list testacl\n")
      output = net_connect.send_command("show hostname\n")
      print " " + line
```
and below is the new formatted code which just keeps giving me errors:

```
from __future__ import absolute_import, division, print_function
import getpass
import sys
import telnetlib
from sys import argv
import csv
import ipaddress
#import request
#from netmiko import ConnectHandler

# print(sys.version)

aclfind, device_lists, = argv

with open(device_lists, 'rb') as f:
    device_lists = [line for line in csv.DictReader(f)]


def get_connection(device):
    device['username'] = "python"
    device['password'] = "cisco123"
    device['secret'] = "cisco123"
    device['device_type'] = 'cisco_ios_ssh'
    try:
        net_connect = ConnectHandler(**device)
    except:  # you should be more explicit with what exceptions you are catching
        try:
            device['device_type'] = 'cisco_ios_telnet'
            net_connect = ConnectHandler(**device)
        except:
            print
            "Unable to connect!"
            net_connect = False
    return net_connect


#net_connect.enable()

for device in device_lists:
    print
    ("\n\n-------------\nDevice: {0} \n---------\n".format(device['host']))
    net_connect = get_connection(device)
    if not net_connect:
        continue

access_lists = net_connect.send_command("show ip access-lists")

if 'access list 101' in access_lists:
    with open('acl_101' + device['host'] + '.txt', 'w') as acl101:
        config = [
            'ip access-list extended 101',
            '109 permit ip 10.10.0.0 0.0.0.255 any',
            'exit',
            'access-list resequence 101',
        ]
        output = net_connect.send_config_set(config)
        net_connect.save_config()
        acl101.write(output + '\n')

if 'access list 102' in access_lists:
    with open('acl_102' + device['host'] + '.txt', 'w') as acl102:
        config = [
            'ip access-list extended 102',
            '109 permit ip 10.10.0.0 0.0.0.255 any',
            'exit',
            'access-list resequence 102',
        ]
        output = net_connect.send_config_set(config)
        net_connect.save_config()
        acl102.write(output + '\n')

if 'access list test' in access_lists:
    with open('acl_test' + device['host'] + '.txt', 'w') as acltest:
        config = [
            'ip access-list extended test',
            '109 permit ip 10.10.0.0 0.0.0.255 any',
            'exit',
            'access-list resequence test',
        ]
        output = net_connect.send_config_set(config)
        net_connect.save_config()
        acltest.write(output + '\n')
```
I totally confused!

**Comment by newscriptkid** [comment]
Hi Gregg,

I've adjusted the python2.7 code based on your recommendations except the 'function' element as once i removed that then the code kind of started working.

The code runs and prints to file but only from one device, even though the code prints the following when i run the script:
```
C:\Program Files\Sublime Text 3>python aclnewcode.py csv_file.csv
Username: python
Password
Enable Password


-------------
Device: R1
---------

Connected via Telnet


-------------
Device: ESW1
---------

Connected via SSH


-------------
Device: ESW2
---------

Connected via Telnet


-------------
Device: ESW3
---------

Connected via Telnet


-------------
Device: R3
---------

Connected via Telnet


Extended IP access list 101
    10 permit ip 172.16.10.0 0.0.0.255 any
Extended IP access list 198
    10 permit ip any any

C:\Program Files\Sublime Text 3>python aclnewcode.py csv_file.csv
Username: python
Password
Enable Password


-------------
Device: R1
---------

Connected via Telnet


-------------
Device: ESW1
---------

Connected via SSH


-------------
Device: ESW2
---------

Connected via Telnet


-------------
Device: ESW3
---------

Connected via Telnet


-------------
Device: R3
---------

Connected via Telnet


Extended IP access list 101
    10 permit ip 172.16.10.0 0.0.0.255 any
    20 permit ip 10.10.0.0 0.0.0.255 any
Extended IP access list 198
    10 permit ip any any

C:\Program Files\Sublime Text 3>
```

What i noticed is it only creates a file for the last device in the csv_file (list of the network devices):
```
host	ip
R1	172.16.10.1
ESW1	192.168.211.131
ESW2	192.168.211.132
ESW3	192.168.176.130
R3	192.168.176.10
```
I tested that by moving the bottom device to the top of the list (R1) and then only R3 tezt file would be created. So i thought lets create a for loop and i did the following:

```
for device in csv_file:
  if 'access list 101' in access_lists:
    print access_lists
    with open('acl_101' + device['host'] + '.txt', 'w') as acl101:
        config = [
            'ip access-list extended 101',
            '109 permit ip 10.10.0.0 0.0.0.255 any',
            'exit',
            'ip access-list resequence 101 10 10',
            'exit',
            'show ip access-list 101',
            'show run | inc hostname',
        ]
        output = net_connect.send_config_set(config)
        net_connect.save_config()
        acl101.write(output + '\n')
        

    if 'access list 102' in access_lists:
      print access_lists
      with open('acl_102' + device['host'] + '.txt', 'w') as acl102:
        config = [
            'ip access-list extended 102',
            '109 permit ip 10.10.0.0 0.0.0.255 any',
            'exit',
            'ip access-list resequence 102 10 10',
            'exit',
            'show ip access-list 101',
            'show run | inc hostname',
        ]
        output = net_connect.send_config_set(config)
        net_connect.save_config()
        acl102.write(output + '\n')

    if 'access list test' in access_lists:
      print access_lists
      with open('acl_test' + device['host'] + '.txt', 'w') as acltest:
        config = [
            'ip access-list extended 101',
            '109 permit ip 10.10.0.0 0.0.0.255 any',
            'exit',
            'ip access-list resequence test 10 10',
            'exit',
            'show ip access-list 101',
            'show run | inc hostname',
        ]
        output = net_connect.send_config_set(config)
        net_connect.save_config()
        acltest.write(output + '\n')
```
But with this it creates 5 files for R3 only (5 files because there are 5 devices and creates 5 files like the example below:

```
acl_101ESW1
acl_101ESW2
acl_101ESW3
acl_101R1
acl_101R3
```
In all the files above it is only R3 running the commands.

**Comment by grelleum** [comment]
It looks like the issue is with your indentation.

The code that connects to the appliance is indented under the for loop:
```
for device in device_lists:
  print "\n\n-------------\nDevice: {0} \n---------\n".format(device['host'])
  try:
    ....
```
So the try/except/connect code will be executed for each device in the list.

From `net_connect.enable()` onward, the code is at the same indentation level as the `for device` loop.  So effectively, none of that code is part of your loop.  So while you connect to all of the devices, you only perform you ACL logic upon that last device in the loop.

```
for device in device_lists:
      # code here is part of the loop due to the indentation.
# not indented - not part of the loop.
net_connect.enable() 
output = net_connect.send_command("show access-list 101")
with open('acl_101' + device['host'] + '.txt', 'w' ) as acl101:
```

Your most recent code also has an indentation problem, such that you will only check for acl102 and acltest if acl101 exists:

```
for device in csv_file:
    if 'access list 101' in access_lists:
        if 'access list 102' in access_lists:
        if 'access list test' in access_lists:
```

Indentation is the way we tell Python which lines of code belong to what code blocks.
It is a recommendation in the Python community to always use four spaces for each level of indentation, that way it is much more obvious that you are properly indenting your code.

**Comment by grelleum** [comment]
Here is a refactored version of your code.  I have not tested it, but it should accomplish what you are looking to do.
```
from __future__ import absolute_import, division, print_function
import getpass
import sys
import telnetlib
from sys import argv
import csv
import ipaddress
import requests
from netmiko import ConnectHandler


def get_connection(device):
    device['username'] = username
    device['password'] = password
    device['secret'] = enablepw
    device['device_type'] = 'cisco_ios_ssh'
    try:
        net_connect = ConnectHandler(**device)
    except:  # you should be more explicit with what exceptions you are catching
        try:
            device['device_type'] = 'cisco_ios_telnet'
            net_connect = ConnectHandler(**device)
        except:
            print("Unable to connect!")
            net_connect = False
    return net_connect


switchdata, device_lists, = argv
with open(device_lists, 'rb') as f:
    device_lists = [line for line in csv.DictReader(f)]

for device in device_lists:
    print("\n\n-------------\nDevice: {0} \n---------\n".format(device['host']))

    # If you want to be prompted only once for all devices
    # then move this before the loop.
    username = raw_input("Username: ")
    password = getpass.getpass("Password ")
    enablepw = getpass.getpass("Enable Password ")

    net_connect = get_connection(device)
    if not net_connect:
        continue

    print("\n")
    net_connect.enable()
    access_lists = net_connect.send_command("show ip access-lists")
    for acl in ['101', '102', 'test']:
        if 'access list ' + acl in access_lists:
            filename = 'acl_{0}{1}.txt'.format(acl, device['host'])
            with open(filename, 'w') as f:
                config = [
                    'ip access-list extended {0}'.format(acl),
                    '109 permit ip 10.10.0.0 0.0.0.255 any',
                    'exit',
                    'access-list resequence {0}'.format(acl),
                ]
                output = net_connect.send_config_set(config)
                net_connect.save_config()
                f.write(output + '\n')
```

**Comment by newscriptkid** [comment]
OMG! 
you are a Super Star!
Please can you explain to me how to think like a scripter/programmer, I just cant find that click in my brain to get a flow with scripting. I have over 12 years experience as a Network Engineer, I have always configured Cisco devices through CLI, and Juniper devices CLI and Gui but i have no issues with CLI. I understand the different cli approaches for juniper and cisco but i just cant find the logic behind the command syntax for python. I kind of map the python logic i have learnt so far to my networking logic but with network devices you have the "?" option to see a list of commands for that level of the config (e.g. interface level).

I really need to find the logic behind the python syntax i think.

Any advice would be highly appreciated.

On a side note I honestly appreciate your help and i didnt expect to get a reply let a lone multiple replies and a solution (what a bonus). I appreciate you must be busy and you have taken time out to help me here.

I have one more project I am working on but I will raise a separate issue for that one, you've already spent more time than i deserved so if you dont get a chance to look at the other issue no worries.

Thanks again,
Shah

**Comment by grelleum** [comment]
Here's the thing.  It took you 12 years to be as proficient with Networking as you are today, probably took at least six months to a year before things started to makes sense, and a lot longer before you were really comfortable with Networking.  The point, of course, is that you can develop the same comfort with Python, it just takes time and it does take practice.  

I recommend trying to write scripts as often as your time permits and always "refactor" your scripts.  Refactoring is re-writing something with the purpose of making it clearer and more readable, without changing the actual behavior.  Many people look at refactoring as not worth the time, if they have something that works, why change it?  You should refactor your code to improve it and make it easier to understand when you want to modify it in the future, and additionally, it heals you to understand how the code works.  The idea is that the first draft is to get the code working, the following rewrites are to make it better.  The more you refactor your code, the less you will need to refactor code that you write.  You will start to naturally write code that is more maintainable.

A year from now, you should expect that the code you write today should "embarrass" you - in that you should look at it and think, "I can't believe I wrote code like that"  - this is the sign that you are growing and learning.

Good Luck!!!

**Comment by newscriptkid** [comment]
100% words of wisdom,
you made me look at this from a totally different angle, i had forgotten the struggles in the early days of my Networking career,

Thank you again for some great advice,
